File: android/src/main/java/com/github/wuxudong/rncharts/charts/ChartGroupHolder.java
package com.github.wuxudong.rncharts.charts;

import com.github.mikephil.charting.charts.BarLineChartBase;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.jobs.ZoomJob;
import com.github.mikephil.charting.utils.MPPointD;
import com.github.mikephil.charting.utils.MPPointF;
import com.github.mikephil.charting.utils.Transformer;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

class ChartHolder {
    public WeakReference<BarLineChartBase> chart;
    public boolean syncX;
    public boolean syncY;

    public ChartHolder(WeakReference<BarLineChartBase> chart, boolean syncX, boolean syncY) {
        this.chart = chart;
        this.syncX = syncX;
        this.syncY = syncY;
    }
}

public class ChartGroupHolder {

    private static Map<String, Map<String, ChartHolder>> chartGroups = new HashMap<>();



    public static synchronized void addChart(String group, String identifier, BarLineChartBase chart, boolean syncX, boolean syncY) {
        if (!chartGroups.containsKey(group)) {
            chartGroups.put(group, new HashMap<String, ChartHolder>());
        }

        chartGroups.get(group).put(identifier, new ChartHolder(new WeakReference<>(chart), syncX, syncY));
    }

    public static synchronized void removeChart(String group, String identifier) {
        if (chartGroups.containsKey(group)) {
            chartGroups.get(group).remove(identifier);
        }
    }

    // sync gesture to other chart in the same group
    public static synchronized void sync(String group, String identifier, float scaleX, float scaleY, float centerX, float centerY) {

        Map<String, ChartHolder> identifierMap = chartGroups.get(group);
        if (identifierMap != null) {
            for (Map.Entry<String, ChartHolder> entry : identifierMap.entrySet()) {
                if (!entry.getKey().equals(identifier)) {
                    ChartHolder chartHolder = entry.getValue();
                    WeakReference<BarLineChartBase> reference = chartHolder.chart;
                    BarLineChartBase chart = reference.get();
                    if (chart != null) {

                        YAxis.AxisDependency axis = chart.getAxisLeft().isEnabled() ? YAxis.AxisDependency.LEFT : YAxis.AxisDependency.RIGHT;
                        Transformer transformer = chart.getTransformer(axis);

                        float zoomScaleX = scaleX / chart.getScaleX();
                        float zoomScaleY = scaleY / chart.getScaleY();

                        MPPointF originalCenter = chart.getCenter();
                        MPPointD originalCenterValue = chart.getValuesByTouchPoint(originalCenter.x, originalCenter.y, axis);

                        float finalScaleX = chartHolder.syncX ? zoomScaleX : 1;
                        float finalScaleY = chartHolder.syncY ? zoomScaleY : 1;

                        float finalCenterX = chartHolder.syncX ? centerX : (float) originalCenterValue.x;
                        float finalCenterY = chartHolder.syncY ? centerY : (float) originalCenterValue.y;

                        ZoomJob.getInstance(chart.getViewPortHandler(), finalScaleX, finalScaleY, finalCenterX, finalCenterY, transformer, axis, chart).run();

                    }
                }
            }
        }
    }
}

File: ios/ReactNativeCharts/ChartGroupHolder.swift
//
//  ChartGroupHolder.swift
//  Charts
//
//  Created by wuxudong on 2018/10/1.
//

import Foundation


class ChartHolder {
    open weak var chart:BarLineChartViewBase?
    open var syncX: Bool;
    open  var syncY:Bool;
    
    init( chart: BarLineChartViewBase,  syncX: Bool,  syncY: Bool) {
        self.chart = chart
        self.syncX = syncX
        self.syncY = syncY
    }
}

open class ChartGroupHolder {
    static var chartGroup: Dictionary<String, Dictionary<String, ChartHolder>> = [:]
    
    public static func addChart(group: String, identifier: String ,  chart: BarLineChartViewBase,  syncX: Bool, syncY: Bool) {
        objc_sync_enter(chartGroup)
        defer { objc_sync_exit(chartGroup) }
        
        let keyExists = chartGroup[group] != nil
        
        if !keyExists {
            chartGroup[group] = [:]
        }
        
        chartGroup[group]![identifier] = ChartHolder(chart: chart, syncX: syncX, syncY: syncY)
    }
    
    
    // sync gesture to other chart in the same group
    public static func sync( group: String,  identifier: String,  scaleX: CGFloat,  scaleY:CGFloat,  centerX: CGFloat,  centerY:CGFloat,  performImmediately: Bool) {
        
        objc_sync_enter(chartGroup)
        defer { objc_sync_exit(chartGroup) }
        
        if let identifierMap = chartGroup[group] {
            for (identifierKey, chartHolder) in identifierMap {
                if identifierKey != identifier {
                    if let chart = chartHolder.chart {
                        
                        let axis = chart.getAxis(YAxis.AxisDependency.left).enabled ? YAxis.AxisDependency.left : YAxis.AxisDependency.right
                        
                        let contentRect = chart.contentRect
                        
                        let originalCenterValue = chart.valueForTouchPoint(point: CGPoint(x: contentRect.midX, y: contentRect.midY), axis: axis)

                        let finalScaleX = chartHolder.syncX ? scaleX : chart.scaleX
                        let finalScaleY = chartHolder.syncY ? scaleY : chart.scaleY
                        
                        let finalCenterX = chartHolder.syncX ? centerX : originalCenterValue.x;
                        let finalCenterY = chartHolder.syncY ? centerY : originalCenterValue.y;
                        
                        chart.zoom(scaleX: finalScaleX, scaleY: finalScaleY, xValue: Double(finalCenterX), yValue: Double(finalCenterY), axis: axis);                        
                    }
                }
            }
        }
    }
}
